src/fold.jl

```diff
@@ -0,0 +1,68 @@
+export unfold, fold
+
+"""
+    unfold(X, W; stride=1, padding=0, dilation=1)
+Extracts sliding local blocks from a batched input tensor. X is the input 5d vector of size 
+`(image_w, image_h, image_d, channels, batch_size)`. Only 3D images are supported. Can be 
+used for 2D or 1D images, by adding an extra singleton dimention. W is the size of kernel, 
+in format `(kernel_w, kernel_h, kernel_d, channels_in, channels_out)`. Note that 
+`channels_in = channels_out`. Output has the size of `(L, channels*kernel_w*kernel_h*kernel_d, batch_size)`, 
+where L is the total number of blocks.
+
+"""
+function unfold(X::AbstractArray{T,5} where T, w_dim::NTuple{5}; stride=1, padding=0, dilation=1)
+    x_dim = size(X)
+    if w_dim[end] != w_dim[end-1]
+        throw(DimensionMismatch("input channels must be equal to output channels in the kernel"))
+    end
+
+    # Make DenseConvDims object
+    cdims = DenseConvDims(x_dim, w_dim; stride=stride, padding=padding, dilation=dilation)
+
+    # Calculate the total number of sliding blocks
+    col_dim = (im2col_dims(cdims))[1:2]
+    col = undef
```

Comment:
I'm a little uncomfortable using undef since it have erratic behaviour in corner cases.

Reply:
Noted. I'll instead use `fill()` to initialise and then use views to fill it with actual values in the loop.